File: server/binance.ts
import { execFileSync } from "node:child_process";
import { ProxyAgent, request } from "undici";
import { daysToExpiry, formatExpiry, parseOptionSymbol } from "../src/lib/options";
import type { OptionContract, OptionSnapshot } from "../src/types";

const BASE_URL = "https://eapi.binance.com";

interface BinanceExchangeInfo {
  serverTime?: number;
  optionSymbols?: Array<{
    symbol: string;
    underlying?: string;
    side?: string;
    expiryDate?: number;
    strikePrice?: string;
    status?: string;
  }>;
}

interface BinanceMark {
  symbol: string;
  markPrice?: string;
}

interface BinanceIndex {
  indexPrice?: string;
}

let lastGoodSnapshot: OptionSnapshot | null = null;

function normalizeProxyUrl(value: string | undefined) {
  if (!value) return null;
  const trimmed = value.trim();
  if (!trimmed) return null;

  const proxyEntry =
    trimmed
      .split(";")
      .map((part) => part.trim())
      .find((part) => part.toLowerCase().startsWith("https=")) ??
    trimmed
      .split(";")
      .map((part) => part.trim())
      .find((part) => part.toLowerCase().startsWith("http=")) ??
    trimmed;

  const proxyValue = proxyEntry.includes("=") ? proxyEntry.split("=").slice(1).join("=") : proxyEntry;
  return /^https?:\/\//i.test(proxyValue) ? proxyValue : `http://${proxyValue}`;
}

function readWindowsProxy() {
  if (process.platform !== "win32") return null;

  try {
    const key = "HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings";
    const proxyEnabled = execFileSync("reg", ["query", key, "/v", "ProxyEnable"], {
      encoding: "utf8",
      timeout: 2000,
      windowsHide: true
    });
    if (!/\b0x1\b/i.test(proxyEnabled)) return null;

    const proxyServer = execFileSync("reg", ["query", key, "/v", "ProxyServer"], {
      encoding: "utf8",
      timeout: 2000,
      windowsHide: true
    });
    const match = proxyServer.match(/ProxyServer\s+REG_SZ\s+(.+)/i);
    return normalizeProxyUrl(match?.[1]);
  } catch {
    return null;
  }
}

function getProxyUrl() {
  return (
    normalizeProxyUrl(process.env.HTTPS_PROXY) ??
    normalizeProxyUrl(process.env.HTTP_PROXY) ??
    normalizeProxyUrl(process.env.ALL_PROXY) ??
    readWindowsProxy()
  );
}

async function requestJson<T>(path: string): Promise<T> {
  const url = `${BASE_URL}${path}`;
  const proxyUrl = getProxyUrl();

  if (proxyUrl) {
    try {
      return await requestJsonWithDispatcher<T>(url, proxyUrl);
    } catch {
      return requestJsonWithDispatcher<T>(url);
    }
  }

  try {
    return await requestJsonWithDispatcher<T>(url);
  } catch (error) {
    throw error;
  }
}

async function requestJsonWithDispatcher<T>(url: string, proxyUrl?: string): Promise<T> {
  const response = await request(url, {
    dispatcher: proxyUrl ? new ProxyAgent(proxyUrl) : undefined,
    bodyTimeout: 30_000,
    headersTimeout: 30_000
  });

  const text = await response.body.text();
  if (response.statusCode < 200 || response.statusCode >= 300) {
    throw new Error(`${url.replace(BASE_URL, "")} ${response.statusCode}`);
  }

  return JSON.parse(text) as T;
}

function toNumber(value: unknown) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

export async function fetchBtcPutSnapshot(nowMs = Date.now()): Promise<OptionSnapshot> {
  try {
    const [exchangeInfo, marks, index] = await Promise.all([
      requestJson<BinanceExchangeInfo>("/eapi/v1/exchangeInfo"),
      requestJson<BinanceMark[]>("/eapi/v1/mark"),
      requestJson<BinanceIndex>("/eapi/v1/index?underlying=BTCUSDT")
    ]);

    const markBySymbol = new Map(marks.map((mark) => [mark.symbol, toNumber(mark.markPrice)]));
    const contracts: OptionContract[] = [];

    for (const item of exchangeInfo.optionSymbols ?? []) {
      const parsed = parseOptionSymbol(item.symbol);
      const expiry = item.expiryDate ?? 0;
      const strike = toNumber(item.strikePrice) ?? parsed?.strike;
      const side = item.side === "PUT" || parsed?.side === "PUT" ? "PUT" : "CALL";

      if (!parsed || parsed.underlying !== "BTC") continue;
      if (side !== "PUT") continue;
      if (!Number.isFinite(expiry) || expiry <= nowMs) continue;
      if (typeof strike !== "number" || !Number.isFinite(strike)) continue;
      if (item.status && item.status !== "TRADING") continue;

      contracts.push({
        symbol: item.symbol,
        expiry,
        expiryLabel: formatExpiry(expiry),
        strike,
        side: "PUT",
        markPrice: markBySymbol.get(item.symbol) ?? null,
        daysToExpiry: daysToExpiry(expiry, nowMs)
      });
    }

    const snapshot: OptionSnapshot = {
      underlying: "BTC",
      indexPrice: toNumber(index.indexPrice),
      serverTime: exchangeInfo.serverTime ?? nowMs,
      updatedAt: nowMs,
      contracts,
      stale: false
    };

    lastGoodSnapshot = snapshot;
    return snapshot;
  } catch (error) {
    const message = error instanceof Error ? error.message : "Binance request failed";
    if (lastGoodSnapshot) {
      return {
        ...lastGoodSnapshot,
        stale: true,
        sourceError: message
      };
    }
    throw new Error(message);
  }
}

File: server/binance.py
from __future__ import annotations

import os
import platform
import time
from datetime import datetime
from typing import Any

import requests

BASE_URL = "https://eapi.binance.com"
DAY_MS = 24 * 60 * 60 * 1000

last_good_snapshot: dict[str, Any] | None = None


def normalize_proxy_url(value: str | None) -> str | None:
    if not value:
        return None

    trimmed = value.strip()
    if not trimmed:
        return None

    parts = [part.strip() for part in trimmed.split(";") if part.strip()]
    proxy_entry = next((part for part in parts if part.lower().startswith("https=")), None)
    proxy_entry = proxy_entry or next((part for part in parts if part.lower().startswith("http=")), None)
    proxy_entry = proxy_entry or trimmed

    proxy_value = proxy_entry.split("=", 1)[1] if "=" in proxy_entry else proxy_entry
    if proxy_value.lower().startswith(("http://", "https://")):
        return proxy_value
    return f"http://{proxy_value}"


def read_windows_proxy() -> str | None:
    if platform.system() != "Windows":
        return None

    try:
        import winreg

        path = r"Software\Microsoft\Windows\CurrentVersion\Internet Settings"
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, path) as key:
            proxy_enabled = winreg.QueryValueEx(key, "ProxyEnable")[0]
            if proxy_enabled != 1:
                return None
            proxy_server = winreg.QueryValueEx(key, "ProxyServer")[0]
            return normalize_proxy_url(str(proxy_server))
    except OSError:
        return None


def get_proxy_url() -> str | None:
    return (
        normalize_proxy_url(os.environ.get("HTTPS_PROXY"))
        or normalize_proxy_url(os.environ.get("HTTP_PROXY"))
        or normalize_proxy_url(os.environ.get("ALL_PROXY"))
        or read_windows_proxy()
    )


def request_json(path: str) -> Any:
    proxy_url = get_proxy_url()
    if proxy_url:
        try:
            return request_json_with_proxy(path, proxy_url)
        except requests.RequestException:
            return request_json_with_proxy(path, None)

    return request_json_with_proxy(path, None)


def request_json_with_proxy(path: str, proxy_url: str | None) -> Any:
    proxies = {"http": proxy_url, "https": proxy_url} if proxy_url else None
    response = requests.get(f"{BASE_URL}{path}", proxies=proxies, timeout=30)
    if not response.ok:
        raise RuntimeError(f"{path} {response.status_code}")
    return response.json()


def to_number(value: Any) -> float | None:
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        return None
    return parsed if parsed == parsed and parsed not in (float("inf"), float("-inf")) else None


def parse_option_symbol(symbol: str) -> dict[str, Any] | None:
    parts = symbol.split("-")
    if len(parts) != 4:
        return None

    underlying, date_part, strike_part, side_part = parts
    strike = to_number(strike_part)
    if not underlying or not date_part or strike is None:
        return None
    if side_part not in ("P", "C"):
        return None

    return {
        "underlying": underlying,
        "datePart": date_part,
        "strike": strike,
        "side": "PUT" if side_part == "P" else "CALL",
    }


def days_to_expiry(expiry_ms: float, now_ms: float) -> float:
    return max(0, (expiry_ms - now_ms) / DAY_MS)


def format_expiry(expiry_ms: float) -> str:
    return datetime.fromtimestamp(expiry_ms / 1000).strftime("%m/%d %H:%M")


def fetch_btc_put_snapshot(now_ms: int | None = None) -> dict[str, Any]:
    global last_good_snapshot

    now = now_ms or int(time.time() * 1000)

    try:
        exchange_info, marks, index = (
            request_json("/eapi/v1/exchangeInfo"),
            request_json("/eapi/v1/mark"),
            request_json("/eapi/v1/index?underlying=BTCUSDT"),
        )

        mark_by_symbol = {item.get("symbol"): to_number(item.get("markPrice")) for item in marks}
        contracts = []

        for item in exchange_info.get("optionSymbols", []):
            symbol = item.get("symbol", "")
            parsed = parse_option_symbol(symbol)
            expiry = item.get("expiryDate") or 0
            strike = to_number(item.get("strikePrice"))
            if strike is None and parsed:
                strike = parsed["strike"]
            side = "PUT" if item.get("side") == "PUT" or (parsed and parsed["side"] == "PUT") else "CALL"

            if not parsed or parsed["underlying"] != "BTC":
                continue
            if side != "PUT":
                continue
            if not isinstance(expiry, (int, float)) or expiry <= now:
                continue
            if strike is None:
                continue
            if item.get("status") and item.get("status") != "TRADING":
                continue

            contracts.append(
                {
                    "symbol": symbol,
                    "expiry": expiry,
                    "expiryLabel": format_expiry(expiry),
                    "strike": strike,
                    "side": "PUT",
                    "markPrice": mark_by_symbol.get(symbol),
                    "daysToExpiry": days_to_expiry(expiry, now),
                }
            )

        snapshot = {
            "underlying": "BTC",
            "indexPrice": to_number(index.get("indexPrice")),
            "serverTime": exchange_info.get("serverTime") or now,
            "updatedAt": now,
            "contracts": contracts,
            "stale": False,
        }
        last_good_snapshot = snapshot
        return snapshot
    except Exception as error:
        message = str(error) or "Binance request failed"
        if last_good_snapshot:
            return {**last_good_snapshot, "stale": True, "sourceError": message}
        raise RuntimeError(message) from error
